archiver/worker.go

```diff
@@ -0,0 +1,264 @@
+		if err := w.markStatus(
+			jobs,
+			jobsdb.Aborted.State,
+			errJson(err),
+		); err != nil {
+			if w.lifecycle.ctx.Err() != nil {
+				return false
+			}
+			log.Errorw("failed to mark unconfigured archive jobs' status", "error", err)
+			panic(err)
+
+		}
+		if !limitReached {
+			return true
+		}
+		goto start
+	}
+	if !storagePrefs.Backup(w.archiveFrom) {
+		if err := w.markStatus(
+			jobs,
+			jobsdb.Aborted.State,
+			errJson(fmt.Errorf("%s archival disabled for workspace %s", w.archiveFrom, workspaceID)),
+		); err != nil {
+			if w.lifecycle.ctx.Err() != nil {
+				return false
+			}
+			log.Errorw("failed to mark archive disabled jobs' status", "error", err)
+			panic(err)
+		}
+		if !limitReached {
+			return true
+		}
+		goto start
+	}
+
+	location, err := w.uploadJobs(w.lifecycle.ctx, jobs)
+	if err != nil {
+		log.Errorw("failed to upload jobs", "error", err)
+		return false
+	}
+	w.lastUploadTime = time.Now()
+
+	if err := w.markStatus(
+		jobs,
+		jobsdb.Succeeded.State,
+		locationJson(location),
+	); err != nil {
+		if w.lifecycle.ctx.Err() != nil {
+			return false
+		}
+		log.Errorw("failed to mark successful upload status", "error", err)
+		panic(err)
+	}
+	if !limitReached {
+		return true
+	}
+	goto start
+}
+
+func (w *worker) SleepDurations() (min, max time.Duration) {
+	if w.lastUploadTime.IsZero() {
+		return w.config.minSleep, w.config.uploadFrequency
+	}
+	return w.config.minSleep, time.Until(w.lastUploadTime.Add(w.config.uploadFrequency))
+}
+
+func (w *worker) Stop() {
+	w.lifecycle.cancel()
+}
+
+func (w *worker) uploadJobs(ctx context.Context, jobs []*jobsdb.JobT) (string, error) {
+	defer w.uploadLimiter.Begin(w.sourceID)()
+	firstJobCreatedAt := jobs[0].CreatedAt.UTC()
+	lastJobCreatedAt := jobs[len(jobs)-1].CreatedAt.UTC()
+	workspaceID := jobs[0].WorkspaceId
+	log := w.log.With("workspaceID", workspaceID)
+
+	gzWriter := fileuploader.NewGzMultiFileWriter()
+	path := fmt.Sprintf(
+		"%v%v.json.gz",
+		lo.Must(misc.CreateTMPDIR())+"/rudder-backups/"+w.sourceID+"/",
+		fmt.Sprintf("%v_%v_%v", firstJobCreatedAt.Unix(), lastJobCreatedAt.Unix(), workspaceID),
+	)
+
+	for _, job := range jobs {
+		j, err := marshalJob(job)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal job: %w", err)
+		}
+		if _, err := gzWriter.Write(path, append(j, '\n')); err != nil {
+			return "", fmt.Errorf("failed to write to file: %w", err)
+		}
+	}
+	if err := gzWriter.Close(); err != nil {
+		return "", fmt.Errorf("failed to close file: %w", err)
+	}
+	defer func() { _ = os.Remove(path) }()
+
+	fileUploader, err := w.storageProvider.GetFileManager(workspaceID)
+	if err != nil {
+		log.Errorw("Skipping storing errors since no file manager is found", "error", err)
+		return "", fmt.Errorf("failed to get file manager: %w", err)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to open file %s: %w", path, err)
+	}
+	defer func() { _ = file.Close() }()
```

Comment:
[blocking] It is not a good idea to ignore the error from `file.Close()`, during close we might need to flush data to disk, and if it fails data will be lost. Given the criticality of those backups, we need to be extra careful

Reply:
We are only opening the file for reading and sending its contents to the object storage. 
If this operation fails, the jobs will not be marked as succeeded and archiver will retry, i.e. a new file will be created containing the same jobs. What should we do in case closing the file that we opened for reading fails here?

Reply:
My bad, due to various comments the order of events was unclear. No a blocking comment, but if `masrhalJob` we don't close the `gzWriter`